app/workers/metrics/record_db_and_es_record_counts_worker.rb

```diff
@@ -0,0 +1,64 @@
+# This worker checks the number of records for each Elasticsearch index and
+# compares it to the number of records in our database.
+#
+# For indexes that don't match a single model (no model or multiple models) we
+# need to implement a custom db_count method on the Search class to do the
+# counts for us
+module Metrics
+  class RecordDbAndEsRecordCountsWorker
+    include Sidekiq::Worker
+
+    sidekiq_options queue: :low_priority, retry: 5
+
+    # Once the Elasticsearch migration is complete we can update this value to
+    # Search::Cluster::SEARCH_CLASSES which will include all indexes.
+    SEARCH_CLASSES = [
+      Search::ChatChannelMembership,
+      Search::ClassifiedListing,
+      Search::Tag,
+    ].freeze
+
+    # Adjustable margin of error - this is how far off the index count can be
+    # from the database count before we raise an error
+    def perform(margin_of_error = 0)
+      SEARCH_CLASSES.each do |search_class|
+        margin_of_error = margin_of_error.to_i
+        db_count = db_count(search_class)
+        index_count = Search::Client.count(index: search_class::INDEX_ALIAS).dig("count")
+        record_difference = (db_count - index_count).abs
+
+        tags = {
+          search_class: search_class,
+          db_count: db_count,
+          index_count: index_count,
+          record_difference: record_difference,
+          margin_of_error: margin_of_error,
+          action: "record_count"
+        }
+
+        if record_difference > margin_of_error
+          tags[:record_count] = "mismatch"
+          DatadogStatsClient.increment("elasticsearch", tags: tags)
+
+          # This will force the job to retry
+          raise DbAndEsRecordsCountMismatch, "#{search_class} record count mismatch"
```

Comment:
Basically, I'm intentionally failing the job here in the case of a mismatch so it tries again.

Reply:
I fear this could lead to a never ending queue of these jobs in Sidekiq. I don't remember what's the algorithm for retries but I guess it's exponential, so if the next tick (the next time this job is scheduled) happens this job is going to be scheduled twice, then maybe three times and if it keeps failing because the delta is enough to trigger the change, you get a never ending cascade of jobs.

I also think you sould run 1 job per class. It will also lower the chances that this job gets stuck and you can have them running independently. Basically add the search class name as an argument and pass it from the rake task.

What do you think?